arangod/VocBase/vocbase.cpp

```diff
@@ -81,6 +97,162 @@
+  explicit VocBaseLogManager(application_features::ApplicationServer& server, DatabaseID database)
+      : _server(server),
+        _logContext(LoggerContext{Logger::REPLICATION2}.with<logContextKeyDatabaseName>(
+            std::move(database))) {}
+
+  [[nodiscard]] auto getReplicatedLogById(replication2::LogId id) const
+      -> std::shared_ptr<replication2::replicated_log::ReplicatedLog const> {
+    auto guard = _guardedData.getLockedGuard();
+    if (auto iter = guard->logs.find(id); iter != guard->logs.end()) {
+      return iter->second;
+    }
+    // THROW_ARANGO_EXCEPTION_PARAMS(TRI_ERROR_REPLICATION_REPLICATED_LOG_NOT_FOUND, id.id());
+    THROW_ARANGO_EXCEPTION_FORMAT(TRI_ERROR_REPLICATION_REPLICATED_LOG_NOT_FOUND,
+                                  "replicated log %" PRIu64 " not found", id.id());
+  }
+
+  [[nodiscard]] auto getReplicatedLogById(replication2::LogId id)
+      -> std::shared_ptr<replication2::replicated_log::ReplicatedLog> {
+    auto guard = _guardedData.getLockedGuard();
+    if (auto iter = guard->logs.find(id); iter != guard->logs.end()) {
+      return iter->second;
+    }
+    // THROW_ARANGO_EXCEPTION_PARAMS(TRI_ERROR_REPLICATION_REPLICATED_LOG_NOT_FOUND, id.id());
+    THROW_ARANGO_EXCEPTION_FORMAT(TRI_ERROR_REPLICATION_REPLICATED_LOG_NOT_FOUND,
+                                  "replicated log %" PRIu64 " not found", id.id());
+  }
+
+  auto resignAll() {
+    auto guard = _guardedData.getLockedGuard();
+    for (auto&& [id, log] : guard->logs) {
+      auto core = log->drop();
+      core.reset();
+    }
+    guard->logs.clear();
+  }
+
+  [[nodiscard]] auto createReplicatedLog(TRI_vocbase_t& vocbase, replication2::LogId id,
+                                         std::optional<std::string> const& collectionName)
+      -> arangodb::ResultT<std::shared_ptr<replication2::replicated_log::ReplicatedLog>> {
+    auto guard = _guardedData.getLockedGuard();
+    return createReplicatedLogOnData(vocbase, id, collectionName, guard.get(),
+                                     _logContext, _server);
+  }
+
+  [[nodiscard]] auto ensureReplicatedLog(TRI_vocbase_t& vocbase, replication2::LogId id,
+                                         std::optional<std::string> const& collectionName)
+      -> arangodb::ResultT<std::shared_ptr<replication2::replicated_log::ReplicatedLog>> {
+    auto guard = _guardedData.getLockedGuard();
+
+    if (auto iter = guard->logs.find(id); iter != guard->logs.end()) {
+      return iter->second;
+    }
+
+    return createReplicatedLogOnData(vocbase, id, collectionName, guard.get(),
+                                     _logContext, _server);
+  }
+
+  [[nodiscard]] auto dropReplicatedLog(TRI_vocbase_t& vocbase,
+                                       arangodb::replication2::LogId id) -> arangodb::Result {
+    LOG_CTX("658c7", DEBUG, _logContext) << "Dropping replicated log " << id;
+
+    StorageEngine& engine = _server.getFeature<EngineSelectorFeature>().engine();
```

Comment:
Same argument as above (not a fan of `getFeature`)

Reply:
I won't fix this for vocbase. This is out of scope for this PR. However, I totally agree with your arguments and I am strongly in favor of introducing more explicit dependencies. In a separate PR. 😁